File: ts/src/samchon/protocol/master/DistributedServer.ts
/// <referece path="../../API.ts" />

/// <reference path="DistributedSystem.ts" />

namespace samchon.protocol.master
{
	export interface IDistributedServer
		extends external.IExternalServer, DistributedSystem
	{
	}

	export abstract class DistributedServer
		extends DistributedSystem
		implements IDistributedServer
	{
		protected ip: string;
		protected port: number;

		public constructor(systemArray: DistributedSystemArray)
		{
			super(systemArray);

			this.ip = "";
			this.port = 0;
		}

		protected abstract createServerConnector(): IServerConnector;

		public connect(): void
		{
			if (this.communicator == null)
				return;

			this.communicator = this.createServerConnector();
			(this.communicator as IServerConnector).connect(this.ip, this.port);
		}

		public getIP(): string
		{
			return this.ip;
		}
		public getPort(): number
		{
			return this.port;
		}
	}
}

File: cpp/samchon/templates/distributed/DistributedServer.cpp
#include <samchon/templates/distributed/DistributedServer.hpp>

#include <samchon/templates/distributed/DistributedSystemArray.hpp>

using namespace samchon::templates::distributed;

DistributedServer::DistributedServer(DistributedSystemArray *systemArray)
	: DistributedSystem(systemArray),
	external::ExternalServer(systemArray)
{
	this->system_array_ = systemArray;
}
DistributedServer::~DistributedServer()
{
}